### PR TITLE
added plotting records functionality

### PR DESCRIPTION
**Pull request by priyatharsan**
More plotting functionality needed to be added.

### CODE REVIEW

File: vulcanai2/models/basenetwork.py [hunk 1]
```diff
@@ -1,17 +1,26 @@
...
 """Defines the basenetwork class"""
+# Core imports
 import abc
 from torch.autograd import Variable
 import torch.nn.modules.loss as loss
+
+# Vulcan imports
 from .layers import *
 from .metrics import Metrics
+from vulcanai2.plotters.visualization import display_record
```
Comment (rfratila) [inline vulcanai2/models/basenetwork.py]: you should do relative imports here

File: vulcanai2/models/basenetwork.py [hunk 2]
```diff
@@ -261,6 +271,14 @@ def get_weights(self):
         """
         return self.state_dict()
 
+    def print_model_structure(self, input_dim=((1, 28, 28))):
```
Comment (rfratila) [inline vulcanai2/models/basenetwork.py]: we shouldn't default the input_dim

File: vulcanai2/models/basenetwork.py [hunk 3]
```diff
@@ -307,18 +326,49 @@ def fit(self, train_loader, val_loader, epochs, retain_graph=None, valid_interv=
...
-                           "Train Loss: {:.6f} | Test Loss: {:.6f} | "
-                           "Train Acc: {:.4f} | Test Acc: {:.4f}".format(
+                           "Train Loss: {:.6f} | Train Acc: {:.4f} in {:.2f}s \n"
+                           "Test Loss: {:.6f} | Test Acc: {:.4f} in {:.2f}s".format(
                                 epoch,
                                 train_loss,
-                                valid_loss,
                                 train_acc,
-                                valid_acc))
+                                train_epoch_time,
```
Comment (rfratila) [inline vulcanai2/models/basenetwork.py]: Why are you also putting in the time if tqdm will tell you how much time is left in that epoch?

File: vulcanai2/plotters/visualization.py [hunk 4]
```diff
@@ -21,350 +20,74 @@
...
-        matplotlib.use('Agg')
-else:
-    matplotlib.use('TkAgg')
 import matplotlib.pyplot as plt
 
 import logging
 logger = logging.getLogger(__name__)
 
 
-#TODO: update to work with pytorch
```
Comment (rfratila) [inline vulcanai2/plotters/visualization.py]: Why did you delete all this stuff? it will need to be implemented. did you move it elsewhere?

File: vulcanai2/models/basenetwork.py (hunk 3 above)
Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: I would prefer to have the epoch results on one line

File: vulcanai2/models/basenetwork.py (hunk 1 above)
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: something like `from ..plotters.visualization import display_record` ?

File: vulcanai2/models/basenetwork.py (hunk 2 above)
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: 👍

File: vulcanai2/models/basenetwork.py (hunk 1 above)
Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: yeah

File: vulcanai2/models/basenetwork.py (hunk 3 above)
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: Makes sense. It was there in Vulcan so I thought I would add it. I will remove

Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: Fair enough. yeah Vulcan's command line info wasn't as robust as tqdm which is why I had it before but this new version works better

File: vulcanai2/plotters/visualization.py (hunk 4 above)
Comment (priyatharsan) [reply vulcanai2/plotters/visualization.py]: We just need to add plotting stuff here. I was working on saliency map got stuck on that the whole day. since I have little knowledge on that. Some functions (`round_list`, `get_confusion_matrix`) are already moved to metrics, some like `get_timestamp` are not necessary in visualization, unless if you'd like them to have?

Comment (rfratila) [reply vulcanai2/plotters/visualization.py]: the most important ones from there are all the functions starting with display (e.g. display_tsne, _plot_reduction, display_pca, display_saliency_map) and stuff like that

**Comment by rfratila** [comment]
When training ends, the graphs disappear. Try to "freeze" them in a way so they don't close once the training terminates

**Comment by priyatharsan** [comment]
yeah right! I will work on that.

File: vulcanai2/models/basenetwork.py [hunk 5]
```diff
@@ -261,6 +271,14 @@ def get_weights(self):
         """
         return self.state_dict()
 
+    def print_model_structure(self, input_dim):
```
Comment (rfratila) [inline vulcanai2/models/basenetwork.py]: actually, since this is a part of basenetwork, do we store the input dimensions anywhere in this class? perhaps we can default to that.

File: vulcanai2/models/basenetwork.py [hunk 6]
```diff
@@ -307,18 +326,47 @@ def fit(self, train_loader, val_loader, epochs, retain_graph=None, valid_interv=
...
+            validation_error=[],
+            validation_accuracy=[]
+        )
+
         try:
+            if plot is True:
+                fig_number = plt.gcf().number + 1 if plt.fignum_exists(1) else 1
+
             for epoch in trange(self.epoch, epochs, desc='Epoch: ', ncols=80):
+                epoch_time = time.time()
```
Comment (rfratila) [inline vulcanai2/models/basenetwork.py]: don't need to calculate these anymore since tqdm does it for us

File: vulcanai2/models/basenetwork.py [hunk 7]
```diff
@@ -307,18 +326,47 @@ def fit(self, train_loader, val_loader, epochs, retain_graph=None, valid_interv=
...
+
                 train_loss, train_acc = self._train_epoch()
+                train_epoch_time = time.time() - epoch_time
```
Comment (rfratila) [inline vulcanai2/models/basenetwork.py]: don't need this either

File: vulcanai2/models/basenetwork.py [hunk 8]
```diff
@@ -307,18 +326,47 @@ def fit(self, train_loader, val_loader, epochs, retain_graph=None, valid_interv=
...
+
                 valid_loss, valid_acc = self._validate()
+                valid_epoch_time = time.time() - epoch_time
```
Comment (rfratila) [inline vulcanai2/models/basenetwork.py]: nor this

File: vulcanai2/plotters/visualization.py [hunk 9]
```diff
@@ -21,18 +20,75 @@
...
-        matplotlib.use('Agg')
-else:
-    matplotlib.use('TkAgg')
 import matplotlib.pyplot as plt
 
 import logging
 logger = logging.getLogger(__name__)
 
 
+def display_record(record=None, load_path=None):
```
Comment (rfratila) [inline vulcanai2/plotters/visualization.py]: lets try to make these a tiny bit prettier. perhaps we can use seaborn?

File: vulcanai2/models/basenetwork.py (hunk 5 above)
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: yeah good idea!

File: vulcanai2/plotters/visualization.py [hunk 10]
```diff
@@ -248,4 +246,31 @@ def display_record(record=None, load_path=None, interactive=True):
 #     sal_map = sal_fun(input_data)
 #     if sal_map.shape != input_data.shape:
 #         raise ValueError('Shape mismatch')
-#     return sal_map
\ No newline at end of file
+#     return sal_map
+
+def Saliency_map(image,model,preprocess,ground_truth,use_gpu=False,method=util.GradType.GUIDED):
```
Comment (rfratila) [inline vulcanai2/plotters/visualization.py]: Saliency map should not be specific to just images. I realize that we only tested it for image based datasets but it will need to be generalized to any incoming datatype. Let's find a better way to send data to this folder. lets have this function signature start with (network, input_x, input_y,...). Im not sure what datatype the variable for `preprocess` is but I think it may stray away from the scope of this function. simply, we just have the input, put it through and get the back gradients, and return those. Let me know if you have any questions

File: vulcanai2/plotters/visualization.py (hunk 4 above)
Comment (priyatharsan) [reply vulcanai2/plotters/visualization.py]: resolved! (partially I guess)

**Comment by priyatharsan** [comment]
merging this PR, feel free to pitch in comments/reviews in this PR. The changes will be made in a new PR.
